File: android-tree-sitter/src/main/java/com/itsaky/androidide/treesitter/TSQuery.java
/*
 *  This file is part of android-tree-sitter.
 *
 *  android-tree-sitter library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  android-tree-sitter library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *  along with android-tree-sitter.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.androidide.treesitter;

import android.text.TextUtils;
import com.itsaky.androidide.treesitter.annotations.GenerateNativeHeaders;
import com.itsaky.androidide.treesitter.util.TSObjectFactoryProvider;
import dalvik.annotation.optimization.FastNative;

public class TSQuery extends TSNativeObject {

  /**
   * An empty query.
   */
  public static final TSQuery EMPTY = new EmptyQuery();

  protected int errorOffset;
  protected int errorType;

  protected String[] captureNames = null;

  /**
   * For internal use only!
   * <p>
   * Constructs an invalid query.
   */
  protected TSQuery(long pointer) {
    super(pointer);
  }

  /**
   * @return Whether the query is valid or not. A {@link TSQuery} is valid if it has a valid pointer
   * to a native object.
   * @deprecated Use {@link TSNativeObject#canAccess()} instead.
   */
  @Deprecated(since = "3.1.0")
  public boolean isValid() {
    return canAccess();
  }

  public int getErrorOffset() {
    return this.errorOffset;
  }

  public TSQueryError getErrorType() {
    return TSQueryError.valueOf(this.errorType);
  }

  /**
   * Get the number of captures in the query.
   *
   * @return The count.
   */
  public int getCaptureCount() {
    checkAccess();
    return Native.captureCount(getNativeObject());
  }

  /**
   * Get the number of patterns in the query.
   *
   * @return The count.
   */
  public int getPatternCount() {
    checkAccess();
    return Native.patternCount(getNativeObject());
  }

  /**
   * Get the number of string literals in the query.
   *
   * @return The count.
   */
  public int getStringCount() {
    checkAccess();
    return Native.stringCount(getNativeObject());
  }

  public String[] getCaptureNames() {
    if (captureNames == null) {
      captureNames = new String[getCaptureCount()];
      for (int i = 0; i < getCaptureCount(); i++) {
        captureNames[i] = getCaptureNameForId(i);
      }
    }
    return captureNames;
  }

  public int getStartByteForPattern(int pattern) {
    checkAccess();
    validatePatternIndex(pattern);
    return Native.startByteForPattern(getNativeObject(), pattern);
  }

  public TSQueryPredicateStep[] getPredicatesForPattern(int pattern) {
    checkAccess();
    validatePatternIndex(pattern);
    return Native.predicatesForPattern(getNativeObject(), pattern);
  }

  public boolean isPatternRooted(int pattern) {
    checkAccess();
    validatePatternIndex(pattern);
    return Native.patternRooted(getNativeObject(), pattern);
  }

  public boolean isPatternNonLocal(int pattern) {
    checkAccess();
    validatePatternIndex(pattern);
    return Native.patternNonLocal(getNativeObject(), pattern);
  }

  public boolean isPatternGuaranteedAtStep(int offset) {
    checkAccess();
    return Native.patternGuaranteedAtStep(getNativeObject(), offset);
  }

  public String getCaptureNameForId(int id) {
    checkAccess();
    return Native.captureNameForId(getNativeObject(), id);
  }

  public String getStringValueForId(int id) {
    checkAccess();
    return Native.stringValueForId(getNativeObject(), id);
  }

  public TSQuantifier getCaptureQuantifierForId(int pattern, int capture) {
    checkAccess();
    validatePatternIndex(pattern);
    return TSQuantifier.forId(Native.captureQuantifierForId(getNativeObject(), pattern, capture));
  }

  @Override
  protected void closeNativeObj() {
    Native.delete(getNativeObject());
  }

  private void validatePatternIndex(int pattern) {
    if (pattern < 0 || pattern >= getPatternCount()) {
      throw new IndexOutOfBoundsException(
        "pattern count: " + getPatternCount() + ", pattern: " + pattern);
    }
  }

  /**
   * Creates a new {@link TSQuery} with the given {@link TSLanguage} and query source. If the
   * language or the query source is invalid, {@link TSQuery#EMPTY} is returned.
   *
   * The query string should contain one or more S-expression patterns. The query is
   * associated with a particular language, and can only be run on syntax nodes parsed with that
   * language.
   *
   * <p>If all of the given patterns are valid, this returns a <code>TSQuery</code>. If a pattern is
   * invalid, this returns <code>null</code>, and provides two pieces of information about the problem:
   * 1. The byte offset of the error is written to the {@link #errorOffset} parameter.
   * 2. The type of error is written to the {@link #errorType} parameter.
   *
   * @param language The {@link TSLanguage} for the query.
   * @param querySource    The query source.
   * @return The {@link TSQuery} object.
   */
  public static TSQuery create(TSLanguage language, String querySource) {
    if (language == null) {
      throw new IllegalArgumentException("Language cannot be null");
    }

    if (querySource == null || TextUtils.getTrimmedLength(querySource) == 0) {
      throw new IllegalArgumentException("Query cannot be null or blank");
    }

    final var query = TSObjectFactoryProvider.getFactory().createQuery(0);
    query.setNativeObject(Native.newQuery(query, language.getNativeObject(), querySource));
    return query;
  }

  /**
   * An empty query. Instances of this class are invalid queries and does not have any patterns,
   * capture names, etc. The <code>get*Count()</code> methods always return <code>0</code>, the
   * <code>get*(int index)</code> methods always throw an {@link IndexOutOfBoundsException} and
   * the <code>get*ForId(int id)</code> methods always throw {@link UnsupportedOperationException}.
   */
  private static final class EmptyQuery extends TSQuery {

    private EmptyQuery() {
      super(0);
    }

    @Override
    public int getCaptureCount() {
      return 0;
    }

    @Override
    public int getPatternCount() {
      return 0;
    }

    @Override
    public int getStringCount() {
      return 0;
    }

    @Override
    public int getErrorOffset() {
      return -1;
    }

    @Override
    public String[] getCaptureNames() {
      return new String[0];
    }

    @Override
    public TSQueryError getErrorType() {
      return TSQueryError.None;
    }

    @Override
    public boolean isPatternGuaranteedAtStep(int offset) {
      return false;
    }

    @Override
    public String getCaptureNameForId(int id) {
      throw new UnsupportedOperationException();
    }

    @Override
    public String getStringValueForId(int id) {
      throw new UnsupportedOperationException();
    }

    @Override
    public void close() {
    }
  }

  @GenerateNativeHeaders(fileName = "query")
  private static class Native {

    static {
      registerNatives();
    }

    @FastNative
    static native void registerNatives();

    @FastNative
    static native long newQuery(TSQuery query, long pointer, String source);

    @FastNative
    static native void delete(long query);

    @FastNative
    static native int captureCount(long query);

    @FastNative
    static native int patternCount(long query);

    @FastNative
    static native int stringCount(long query);

    @FastNative
    static native int startByteForPattern(long query, int pattern);

    @FastNative
    static native TSQueryPredicateStep[] predicatesForPattern(long query, int pattern);

    @FastNative
    static native boolean patternRooted(long query, int pattern);

    @FastNative
    static native boolean patternNonLocal(long query, int pattern);

    @FastNative
    static native boolean patternGuaranteedAtStep(long query, int byteOffset);

    @FastNative
    static native String captureNameForId(long query, int id);

    @FastNative
    static native String stringValueForId(long query, int id);

    @FastNative
    static native int captureQuantifierForId(long query, int pattern, int capture);
  }
}

File: android-tree-sitter/src/main/cpp/ts_query.cc
/*
 *  This file is part of android-tree-sitter.
 *
 *  android-tree-sitter library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  android-tree-sitter library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *  along with android-tree-sitter.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "utils/ts_exceptions.h"
#include "utils/ts_obj_utils.h"
#include "utils/ts_preconditions.h"

#include "ts_query.h"

void fillQuery(JNIEnv *, jobject, uint32_t, TSQueryError);
int query_quantifier_id(JNIEnv *env, TSQuantifier quantifier);
jint getErrorType(TSQueryError);

static jlong TSQuery_newQuery(JNIEnv *env,
                              jclass self,
                              jobject queryObject,
                              jlong language,
                              jstring source) {
  req_nnp(env, language);
  const char *c_source;
  uint32_t source_length = env->GetStringLength(source);
  c_source = env->GetStringUTFChars(source, nullptr);
  auto error_offset = new uint32_t;
  auto error_type = new TSQueryError;
  TSQuery *query = ts_query_new((TSLanguage *) language,
                                c_source,
                                source_length,
                                error_offset,
                                error_type);
  fillQuery(env, queryObject, *error_offset, *error_type);
  env->ReleaseStringUTFChars(source, c_source);
  return (jlong) query;
}

static void TSQuery_delete(JNIEnv *env, jclass self, jlong query) {
  req_nnp(env, query);
  ts_query_delete((TSQuery *) query);
}

static jint TSQuery_captureCount(JNIEnv *env, jclass self, jlong query) {
  req_nnp(env, query);
  return (jint) ts_query_capture_count((TSQuery *) query);
}

static jint TSQuery_patternCount(JNIEnv *env, jclass self, jlong query) {
  req_nnp(env, query);
  return (jint) ts_query_pattern_count((TSQuery *) query);
}

static jint TSQuery_stringCount(JNIEnv *env, jclass self, jlong query) {
  req_nnp(env, query);
  return (jint) ts_query_string_count((TSQuery *) query);
}

static jint TSQuery_startByteForPattern(JNIEnv *env,
                                        jclass self,
                                        jlong query,
                                        jint pattern) {
  req_nnp(env, query);
  return (jint) ts_query_start_byte_for_pattern((TSQuery *) query, pattern);
}

static jobjectArray TSQuery_predicatesForPattern(JNIEnv *env,
                                                 jclass self,
                                                 jlong query,
                                                 jint pattern) {
  req_nnp(env, query);

  uint32_t count;
  const TSQueryPredicateStep *predicates =
      ts_query_predicates_for_pattern((TSQuery *) query, pattern, &count);
  jobjectArray result = createQueryPredicateStepArr(env, (jint) count);
  req_nnp(env, result, "TSQueryPredicateStep[] from factory");

  for (uint32_t i = 0; i < count; i++) {
    const TSQueryPredicateStep *predicate = (predicates + i);
    jobject obj = _marshalQueryPredicateStep(env, predicate);
    env->SetObjectArrayElement(result, i, obj);
  }

  return result;
}

static jboolean
TSQuery_patternRooted(JNIEnv *env, jclass self, jlong query, jint pattern) {
  req_nnp(env, query);
  return (jboolean) ts_query_is_pattern_rooted((TSQuery *) query, pattern);
}

static jboolean TSQuery_patternNonLocal(JNIEnv *env,
                                        jclass self,
                                        jlong query,
                                        jint pattern_index) {
  req_nnp(env, query);
  return (jboolean) ts_query_is_pattern_non_local((TSQuery *) query,
                                                  pattern_index);
}

static jboolean TSQuery_patternGuaranteedAtStep(JNIEnv *env,
                                                jclass self,
                                                jlong query,
                                                jint offset) {
  req_nnp(env, query);
  return (jboolean) ts_query_is_pattern_guaranteed_at_step((TSQuery *) query,
                                                           offset);
}

static jstring
TSQuery_captureNameForId(JNIEnv *env, jclass self, jlong query, jint id) {
  req_nnp(env, query);
  uint32_t count;
  const char
      *name = ts_query_capture_name_for_id((TSQuery *) query, id, &count);
  return (jstring) env->NewStringUTF(name);
}

static jstring
TSQuery_stringValueForId(JNIEnv *env, jclass self, jlong query, jint id) {
  req_nnp(env, query);
  uint32_t count;
  const char *str = ts_query_string_value_for_id((TSQuery *) query, id, &count);
  return env->NewStringUTF(str);
}

void fillQuery(JNIEnv *env,
               jobject query,
               uint32_t error_offset,
               TSQueryError error_type) {
  req_nnp(env, query);
  jclass klass = env->GetObjectClass(query);
  jfieldID offsetField = env->GetFieldID(klass, "errorOffset", "I");
  jfieldID errorTypeField = env->GetFieldID(klass, "errorType", "I");
  env->SetIntField(query, offsetField, (jint) error_offset);
  env->SetIntField(query, errorTypeField, getErrorType(error_type));
}

jint getErrorType(TSQueryError error) {
  switch (error) {
    default:
    case TSQueryErrorNone:
      return 0;
    case TSQueryErrorSyntax:
      return 1;
    case TSQueryErrorNodeType:
      return 2;
    case TSQueryErrorField:
      return 3;
    case TSQueryErrorCapture:
      return 4;
    case TSQueryErrorStructure:
      return 5;
    case TSQueryErrorLanguage:
      return 6;
  }
}
static jint TSQuery_captureQuantifierForId(JNIEnv *env,
                                           jclass clazz,
                                           jlong query,
                                           jint pattern,
                                           jint capture) {
  req_nnp(env, query);
  auto ts_query = (TSQuery *) query;
  auto quantifier =
      ts_query_capture_quantifier_for_id(ts_query, pattern, capture);
  return query_quantifier_id(env, quantifier);
}

int query_quantifier_id(JNIEnv *env, TSQuantifier quantifier) {
  switch (quantifier) {
    case TSQuantifierZero:
      return 0;
    case TSQuantifierZeroOrOne:
      return 1;
    case TSQuantifierZeroOrMore:
      return 2;
    case TSQuantifierOne:
      return 3;
    case TSQuantifierOneOrMore:
      return 4;
  }

  throw_illegal_args(env, "Unknown quantifier");
  return -1;
}

void TSQuery_Native__SetJniMethods(JNINativeMethod *methods, int count) {
  SET_JNI_METHOD(methods, TSQuery_Native_newQuery, TSQuery_newQuery);
  SET_JNI_METHOD(methods, TSQuery_Native_delete, TSQuery_delete);
  SET_JNI_METHOD(methods, TSQuery_Native_captureCount, TSQuery_captureCount);
  SET_JNI_METHOD(methods, TSQuery_Native_patternCount, TSQuery_patternCount);
  SET_JNI_METHOD(methods, TSQuery_Native_stringCount, TSQuery_stringCount);
  SET_JNI_METHOD(methods, TSQuery_Native_startByteForPattern,
                 TSQuery_startByteForPattern);
  SET_JNI_METHOD(methods, TSQuery_Native_predicatesForPattern,
                 TSQuery_predicatesForPattern);
  SET_JNI_METHOD(methods, TSQuery_Native_patternRooted, TSQuery_patternRooted);
  SET_JNI_METHOD(methods, TSQuery_Native_patternNonLocal, TSQuery_patternNonLocal);
  SET_JNI_METHOD(methods, TSQuery_Native_patternGuaranteedAtStep,
                 TSQuery_patternGuaranteedAtStep);
  SET_JNI_METHOD(methods, TSQuery_Native_captureNameForId, TSQuery_captureNameForId);
  SET_JNI_METHOD(methods, TSQuery_Native_stringValueForId, TSQuery_stringValueForId);
  SET_JNI_METHOD(methods, TSQuery_Native_captureQuantifierForId,
                 TSQuery_captureQuantifierForId);
}